### ISSUE TITLE
StagingBelt write textures

### ISSUE THREAD

**Issue by siriux** [gfx-rs/wgpu-rs#721]
When using the StagingBelt to upload textures currently you need to upload first to an intermediate buffer on the GPU using the belt, and then copy from this buffer to the texture using copy_buffer_to_texture.

Wouldn't it make sense to have a write_texture method on the StagingBelt that performs the copy_buffer_to_texture directly?

Maybe there is a hidden benefit of the intermediate buffer I'm missing, and in this case I would appreciate an explanation of the recommended way to upload frequent texture changes.

Thanks

**Comment by kvark** [comment]
> Wouldn't it make sense to have a write_texture method on the StagingBelt that performs the copy_buffer_to_texture directly?

Yes, absolutely, it's just something we omitted originally. The idea is to have a user library analogue that covers both `write_buffer` and `write_texture`. Any help is appreciated!

**Comment by dhardy** [comment]
I was thinking about this... `COPY_BYTES_PER_ROW_ALIGNMENT = 256` which is significantly larger than some images (e.g. glyphs are only a few pixels wide), hence it might make sense to use a somewhat different approach:

- use a number of 256x256 "chunks"
- use a fast texture atlas (e.g. [`guillotiere::SimpleAtlasAllocator`](https://docs.rs/guillotiere/0.6.0/guillotiere/struct.SimpleAtlasAllocator.html)) to allocate images within the chunk, clearing the allocator after upload
- automatically split large image uploads into chunks
- queue a `copy_buffer_to_texure` for each image (part)

The downside is that it's a little complex and potentially adds dependencies. I also don't know if perf. benefits would be worthwhile over simply creating a new buffer for each upload (though for many small glyphs/icons the alignment restriction implies a lot of wasted space).

**Comment by dhardy** [comment]
[`etagere::BucketedAtlasAllocator`](https://docs.rs/etagere/0.2.4/etagere/struct.BucketedAtlasAllocator.html) might be better... not sure. Potentially we could copy the code, stripping out anything not needed (like deallocation). @nical I think you would be okay with this (at least, the licence permits it)?

**Comment by nical** [comment]
Of course. My recommendation since you don't need deallocation is to use guillotiere::SimpleAtlasAllocator's algorithm, with allocations sorted from largest to smallest (as a rule of thumb).

**Comment by kvark** [comment]
@dhardy if we are talking about any integration with other crates, like `etagere`, then it's no longer something we want in wgpu-rs utils. Let other crates take care of this, and you are welcome to write one!

**Comment by dhardy** [comment]
@kvark as mentioned above, the necessary code could be copied into `wgpu-rs`. I'm guessing it would weigh a couple hundred lines.

There are simpler algorithms for allocating rects within a texture of course, but I see little point in using home-made half-measures when a much better allocator can be copied from a purpose-built tool at very little cost.

**Comment by kvark** [comment]
I feel like texture packing is too complex of an area to try exposing, even if it's just a couple hundred of LOCs. There are many ways to do it, and we don't want to be opinionated in the utils.

**Comment by dhardy** [comment]
My 2 cents is that an efficient solution has to involve some creativity. Alternatives are:

- no packing/sharing at all — e.g. use a 256x10 buffer to upload an 8x10 texture (to meet align width)
- use a simple home-grown allocator, e.g. just advance the x coordinate for each texture until full

**Comment by kvark** [comment]
Maybe this needs to be more generic then, like "give me a staging area of size X", which you'd follow by encoding the copy yourself. In this case, you can combine multiple updates (by packing them into the different offsets of 256-byte chunks for smaller textures).

**Comment by dhardy** [comment]
That doesn't fit with the way I'd want to use the staging belt: let me just push the next glyph/icon and let the staging belt figure out how to upload it in time for the next frame. (That's what it does for buffers, right?)

**Comment by kvark** [comment]
That's what `Queue::write_texture` does for you, it's most efficient. Unfortunately, we can't provide `StagingBelt::write_texture` with the same level of efficiency, because it's purely user side. You'd have a choice between:
  1. eat the cost of having larger staging buffers (for 256 byte alignment)
  2. roll in your own texture packing solution on top of wgpu-rs (potentially re-implementing parts of the staging belt)
  3. switch to `Queue::write_texture`

**Comment by dhardy** [comment]
Yes, `StagingBelt` is purely user-side and could be pushed to a utility crate. A "staging belt for textures" could be entirely separate and in a utility crate, though it would also be possible to use the same buffers to copy to target buffers and textures.

 Lets summarise what `StagingBelt` currently does:
```rust
impl StagingBelt {
    pub fn new(chunk_size: BufferAddress) -> Self;

    pub fn write_buffer(
        &mut self,
        encoder: &mut CommandEncoder,
        target: &Buffer,
        offset: BufferAddress,
        size: BufferSize,
        device: &Device,
    ) -> BufferViewMut;
```
The method `write_buffer`:

- checks whether the copy will fit in any active buffer, after current content (1D allocator)
- otherwise, whether it fits in any free buffer
- otherwise allocates a buffer with size `max(chunk_size, size)`
- queues a copy from the buffer to the `target`
- lets the user fill the buffer via mut slice

There is one slightly questionable decision here: using `max(chunk_size, size)` instead of forcing the user to chunk large writes. But doing that would force a worse API on users, so it is a reasonable choice.

In comparison, a hypothetical `write_texture` method:

- has a 2D allocation problem, to which there are a range of solutions, as mentioned above
- could probably expect the user pass a data slice *in* instead of returning a mut slice *out*

So we could:

1. implement an (a) efficient or (b) naive 2D allocator in crate (my original suggestion) and call it done
2. make `StagingBelt` generic over an allocator trait of our own design and provide a naive implementation as the default type parameter (pushing the "efficient allocation" problem to the user, at some cost in complexity)
3. close this issue with the decision not to add `StagingBelt::write_texture`

Personally I still favour the first option but care less whether we spend a couple hundred lines on an efficient allocator or just a couple on a naive shelf allocator. If you pick (1 a/b) or (2) I may be generous enough to write an implementation.

**Comment by kvark** [comment]
> There is one slightly questionable decision here: using max(chunk_size, size) instead of forcing the user to chunk large writes.

Agreed. The idea was for this to be an alternative of the built-in `write_buffer`, so we didn't want to enforce more constraints.

> Personally I still favour the first option but care less whether we spend a couple hundred lines on an efficient allocator or just a couple on a naive shelf allocator. If you pick (1 a/b) or (2) I may be generous enough to write an implementation.

I think the staging belt is complex enough (and opinionated!) that it deserves to be in its own crate. That would allow it to hook up the 2D allocator dependency and be more aggressive in the optimization.

I know @cwfitzgerald has their own version of the belt, and maybe other folks do as well. Would there be a chance to re-converge the implementations into one? I'd be happy to remove the code from wgpu-rs utils and point to the crate from README.
